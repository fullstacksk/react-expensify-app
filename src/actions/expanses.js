
import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';
// ADD_EXPANSE
export const addExpanse = (expanse) => (
    {
        type: 'ADD_EXPANSE',
        expanse
    }
);
export const startAddExpanse = (expanseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expanseData;
        const expanse = { description, note, amount, createdAt };
        return database.ref('expanses')
            .push(expanse)
            .then((ref) => {
                dispatch(addExpanse({
                    id: ref.key,
                    ...expanse
                }))
            })
            .catch((e) => {
                console.log("Add Expanse Failed", e)
            })
    }
}

// REMOVE_EXPANSE
export const removeExpanse = ({ id = '' } = {}) => ({
    type: 'REMOVE_EXPANSE',
    id

});
// EDIT_EXPANSE
export const editExpanse = (id, updates) => ({
    type: 'EDIT_EXPANSE',
    id,
    updates
});