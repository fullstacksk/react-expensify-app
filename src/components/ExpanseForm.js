import React from 'react';
import moment from 'moment';
import Form from 'react-bootstrap/Form';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

export default class ExpanseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expanse ? props.expanse.id : '',
            description: props.expanse ? props.expanse.description : "",
            note: props.expanse ? props.expanse.note : "",
            amount: props.expanse ? props.expanse.amount.toString() : "",
            createdAt: props.expanse ? moment(props.expanse.craetedAt) : moment(),
            calenderFocused: false,
            error: ''
        }
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Please provide description and amount!!" }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {

        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
            this.setState(() => ({ amount }))
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    }
    render() {
        return (
            <Container>
                <Row className="justify-content-md-center">
                    <Col md={6}>
                        {
                            this.state.error &&
                            <Alert variant="warning">
                                {this.state.error}
                            </Alert>
                        }
                        <Form onSubmit={this.onSubmit} >
                            <Form.Group >
                                <Form.Label>Description</Form.Label>
                                <Form.Control
                                    type="text"
                                    id="description"
                                    value={this.state.description}
                                    onChange={this.onDescriptionChange}
                                />
                            </Form.Group>
                            <Form.Group >
                                <Form.Label>Note</Form.Label>
                                <Form.Control
                                    as="textarea"
                                    rows="2"
                                    id="note"
                                    value={this.state.note}
                                    onChange={this.onNoteChange}
                                />
                            </Form.Group>
                            <Form.Group as={Row}>
                                <Col>
                                    <Form.Label>Amount</Form.Label>
                                    <Form.Control
                                        type="text"
                                        id="amount"
                                        value={this.state.amount}
                                        onChange={this.onAmountChange}
                                    />
                                </Col>
                                <Col id="singleDatePicker">
                                    <Form.Label>Created At</Form.Label>
                                    <br />
                                    <SingleDatePicker
                                        date={this.state.createdAt}
                                        onDateChange={this.onDateChange}
                                        focused={this.state.calenderFocused}
                                        onFocusChange={this.onFocusChange}
                                        numberOfMonths={1}
                                        isOutsideRange={() => false}
                                        anchorDirection="right"
                                    />
                                </Col>
                            </Form.Group>
                            <Form.Group as={Row}>
                                <Col sm={6} className="mb-1">
                                    <Button variant="info" className="w-100" type="submit">Add Expanse </Button>
                                </Col>
                                {this.props.expanse &&
                                    <Col sm={6}>
                                        <Button
                                            id="removeExpanse"
                                            variant="success"
                                            className="w-100"
                                            onClick={this.props.startRemoveExpanse}
                                        >Remove Expanse </Button>
                                    </Col>}
                            </Form.Group>
                        </Form>
                    </Col>
                </Row>
            </Container>
        )
    }
}
