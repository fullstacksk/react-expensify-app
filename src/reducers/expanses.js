
//Expanses Reducer
const expansesReducerDefaultState = [];

export default ((state = expansesReducerDefaultState, action) => {
    switch (action.type) {
        case "SET_EXPANSES":
            return action.expanses
        case 'ADD_EXPANSE':
            return [
                ...state,
                action.expanse
            ];
        case 'REMOVE_EXPANSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPANSE':
            return state.map((expanse) => {
                if (expanse.id === action.id)
                    return {
                        ...expanse,
                        ...action.updates
                    }
                return expanse;
            });
        default:
            return state;
    }
});
