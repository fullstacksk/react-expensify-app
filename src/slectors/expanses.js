import moment from 'moment';
//get visible expanses
export default (expanses, { text, sortBy, startDate, endDate }) => {
    return expanses.filter((expanse) => {
        const craetedAtMoment = moment(expanse.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(craetedAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(craetedAtMoment, 'day') : true;
        const textMatch = expanse.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            // console.log(a, b);
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    })
}
