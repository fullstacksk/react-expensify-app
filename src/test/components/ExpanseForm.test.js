import React from 'react'
import moment from 'moment'
import ExpanseForm from '../../components/ExpanseForm'
import { shallow } from 'enzyme'
import expanses from '../fixtures/expanses'

test("Should render ExpanseForm correctly", () => {
    const wrapper = shallow(<ExpanseForm />)
    expect(wrapper).toMatchSnapshot()
})

test("Should render ExpanseForm with expanse data", () => {
    const wrapper = shallow(<ExpanseForm expanse={expanses[0]} />)
    expect(wrapper).toMatchSnapshot()
})

test("Should render ExpanseForm with error on invalid form submission", () => {
    const wrapper = shallow(<ExpanseForm />)
    expect(wrapper).toMatchSnapshot()
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(wrapper.state('error').length).toBeGreaterThan(0)
    expect(wrapper).toMatchSnapshot()
})

test("Should set description on input change", () => {
    const value = "New Description"
    const wrapper = shallow(<ExpanseForm />)
    // console.log(wrapper.find('Form').closest('Form.Control'))
    wrapper.find('#description').simulate('change', {
        target: { value }
    })
    expect(wrapper.state('description')).toBe(value)
})

test("Should set note on textarea change", () => {
    const value = "New Note"
    const wrapper = shallow(<ExpanseForm />)
    // console.log(wrapper.find('Form').closest('Form.Control'))
    wrapper.find('#note').simulate('change', {
        target: { value }
    })
    expect(wrapper.state('note')).toBe(value)
})

test("Should set amount if valid input", () => {
    const value = '22.34'
    const wrapper = shallow(<ExpanseForm />)
    // console.log(wrapper.find('Form').closest('Form.Control'))
    wrapper.find('#amount').simulate('change', {
        target: { value }
    })
    expect(wrapper.state('amount')).toBe(value)
})

test("Should not set amount if invalid input", () => {
    const value = '22.a34'
    const wrapper = shallow(<ExpanseForm />)
    // console.log(wrapper.find('Form').closest('Form.Control'))
    wrapper.find('#amount').simulate('change', {
        target: { value }
    })
    expect(wrapper.state('amount')).toBe("")
})


test("Should call onSubmit prop for valid form submission", () => {
    const onSubmitSpy = jest.fn()
    const wrapper = shallow(<ExpanseForm expanse={expanses[0]} onSubmit={onSubmitSpy} />)
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(wrapper.state('error')).toBe('')
    expect(onSubmitSpy).toHaveBeenCalledWith({
        description: expanses[0].description,
        note: expanses[0].note,
        amount: expanses[0].amount,
        createdAt: expanses[0].createdAt
    })
})

test("Should set new date onDateChange", () => {
    const now = moment()
    const wrapper = shallow(<ExpanseForm />)
    wrapper.find('#singleDatePicker').childAt(2).prop('onDateChange')(now)
    expect(wrapper.state('createdAt')).toEqual(now)
})

test("Should set calenderFocused onFocusChange", () => {
    const focused = true
    const wrapper = shallow(<ExpanseForm />)
    wrapper.find('#singleDatePicker').childAt(2).prop('onFocusChange')({ focused })
    expect(wrapper.state('calenderFocused')).toBe(focused)
})