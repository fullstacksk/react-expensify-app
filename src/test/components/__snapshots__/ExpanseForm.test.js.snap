// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpanseForm correctly 1`] = `
<Container
  fluid={false}
>
  <Row
    className="justify-content-md-center"
    noGutters={false}
  >
    <Col
      md={6}
    >
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Description
          </FormLabel>
          <FormControl
            id="description"
            onChange={[Function]}
            type="text"
            value=""
          />
        </FormGroup>
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Note
          </FormLabel>
          <FormControl
            as="textarea"
            id="note"
            onChange={[Function]}
            rows="2"
            value=""
          />
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col>
            <FormLabel
              column={false}
              srOnly={false}
            >
              Amount
            </FormLabel>
            <FormControl
              id="amount"
              onChange={[Function]}
              type="text"
              value=""
            />
          </Col>
          <Col
            id="singleDatePicker"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Created At
            </FormLabel>
            <br />
            <withStyles(SingleDatePicker)
              anchorDirection="right"
              appendToBody={false}
              block={false}
              calendarInfoPosition="bottom"
              customCloseIcon={null}
              customInputIcon={null}
              date={"1970-01-01T00:00:00.000Z"}
              dayPickerNavigationInlineStyles={null}
              daySize={39}
              disableScroll={false}
              disabled={false}
              displayFormat={[Function]}
              enableOutsideDays={false}
              firstDayOfWeek={null}
              focused={false}
              hideKeyboardShortcutsPanel={false}
              horizontalMargin={0}
              horizontalMonthPadding={13}
              id="date"
              initialVisibleMonth={null}
              inputIconPosition="before"
              isDayBlocked={[Function]}
              isDayHighlighted={[Function]}
              isOutsideRange={[Function]}
              isRTL={false}
              keepFocusOnInput={false}
              keepOpenOnDateSelect={false}
              monthFormat="MMMM YYYY"
              navNext={null}
              navPosition="navPositionTop"
              navPrev={null}
              noBorder={false}
              numberOfMonths={1}
              onClose={[Function]}
              onDateChange={[Function]}
              onFocusChange={[Function]}
              onNextMonthClick={[Function]}
              onPrevMonthClick={[Function]}
              openDirection="down"
              orientation="horizontal"
              phrases={
                Object {
                  "calendarLabel": "Calendar",
                  "chooseAvailableDate": [Function],
                  "clearDate": "Clear Date",
                  "closeDatePicker": "Close",
                  "dateIsSelected": [Function],
                  "dateIsUnavailable": [Function],
                  "enterKey": "Enter key",
                  "escape": "Escape key",
                  "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                  "homeEnd": "Home and end keys",
                  "jumpToNextMonth": "Move forward to switch to the next month.",
                  "jumpToPrevMonth": "Move backward to switch to the previous month.",
                  "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardShortcuts": "Keyboard Shortcuts",
                  "leftArrowRightArrow": "Right and left arrow keys",
                  "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                  "moveFocusByOneMonth": "Switch months.",
                  "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                  "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                  "openThisPanel": "Open this panel.",
                  "pageUpPageDown": "page up and page down keys",
                  "questionMark": "Question mark",
                  "returnFocusToInput": "Return to the date input field.",
                  "roleDescription": "datepicker",
                  "selectFocusedDate": "Select the date in focus.",
                  "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                  "upArrowDownArrow": "up and down arrow keys",
                }
              }
              placeholder="Date"
              readOnly={false}
              regular={false}
              renderCalendarInfo={null}
              renderDayContents={null}
              renderMonthElement={null}
              renderMonthText={null}
              renderNavNextButton={null}
              renderNavPrevButton={null}
              renderWeekHeaderElement={null}
              reopenPickerOnClearDate={false}
              required={false}
              screenReaderInputMessage=""
              showClearDate={false}
              showDefaultInputIcon={false}
              small={false}
              verticalHeight={null}
              verticalSpacing={22}
              weekDayFormat="dd"
              withFullScreenPortal={false}
              withPortal={false}
            />
          </Col>
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col
            className="mb-1"
            sm={6}
          >
            <Button
              active={false}
              className="w-100"
              disabled={false}
              type="submit"
              variant="info"
            >
              Add Expanse 
            </Button>
          </Col>
        </FormGroup>
      </Form>
    </Col>
  </Row>
</Container>
`;

exports[`Should render ExpanseForm with error on invalid form submission 1`] = `
<Container
  fluid={false}
>
  <Row
    className="justify-content-md-center"
    noGutters={false}
  >
    <Col
      md={6}
    >
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Description
          </FormLabel>
          <FormControl
            id="description"
            onChange={[Function]}
            type="text"
            value=""
          />
        </FormGroup>
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Note
          </FormLabel>
          <FormControl
            as="textarea"
            id="note"
            onChange={[Function]}
            rows="2"
            value=""
          />
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col>
            <FormLabel
              column={false}
              srOnly={false}
            >
              Amount
            </FormLabel>
            <FormControl
              id="amount"
              onChange={[Function]}
              type="text"
              value=""
            />
          </Col>
          <Col
            id="singleDatePicker"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Created At
            </FormLabel>
            <br />
            <withStyles(SingleDatePicker)
              anchorDirection="right"
              appendToBody={false}
              block={false}
              calendarInfoPosition="bottom"
              customCloseIcon={null}
              customInputIcon={null}
              date={"1970-01-01T00:00:00.000Z"}
              dayPickerNavigationInlineStyles={null}
              daySize={39}
              disableScroll={false}
              disabled={false}
              displayFormat={[Function]}
              enableOutsideDays={false}
              firstDayOfWeek={null}
              focused={false}
              hideKeyboardShortcutsPanel={false}
              horizontalMargin={0}
              horizontalMonthPadding={13}
              id="date"
              initialVisibleMonth={null}
              inputIconPosition="before"
              isDayBlocked={[Function]}
              isDayHighlighted={[Function]}
              isOutsideRange={[Function]}
              isRTL={false}
              keepFocusOnInput={false}
              keepOpenOnDateSelect={false}
              monthFormat="MMMM YYYY"
              navNext={null}
              navPosition="navPositionTop"
              navPrev={null}
              noBorder={false}
              numberOfMonths={1}
              onClose={[Function]}
              onDateChange={[Function]}
              onFocusChange={[Function]}
              onNextMonthClick={[Function]}
              onPrevMonthClick={[Function]}
              openDirection="down"
              orientation="horizontal"
              phrases={
                Object {
                  "calendarLabel": "Calendar",
                  "chooseAvailableDate": [Function],
                  "clearDate": "Clear Date",
                  "closeDatePicker": "Close",
                  "dateIsSelected": [Function],
                  "dateIsUnavailable": [Function],
                  "enterKey": "Enter key",
                  "escape": "Escape key",
                  "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                  "homeEnd": "Home and end keys",
                  "jumpToNextMonth": "Move forward to switch to the next month.",
                  "jumpToPrevMonth": "Move backward to switch to the previous month.",
                  "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardShortcuts": "Keyboard Shortcuts",
                  "leftArrowRightArrow": "Right and left arrow keys",
                  "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                  "moveFocusByOneMonth": "Switch months.",
                  "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                  "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                  "openThisPanel": "Open this panel.",
                  "pageUpPageDown": "page up and page down keys",
                  "questionMark": "Question mark",
                  "returnFocusToInput": "Return to the date input field.",
                  "roleDescription": "datepicker",
                  "selectFocusedDate": "Select the date in focus.",
                  "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                  "upArrowDownArrow": "up and down arrow keys",
                }
              }
              placeholder="Date"
              readOnly={false}
              regular={false}
              renderCalendarInfo={null}
              renderDayContents={null}
              renderMonthElement={null}
              renderMonthText={null}
              renderNavNextButton={null}
              renderNavPrevButton={null}
              renderWeekHeaderElement={null}
              reopenPickerOnClearDate={false}
              required={false}
              screenReaderInputMessage=""
              showClearDate={false}
              showDefaultInputIcon={false}
              small={false}
              verticalHeight={null}
              verticalSpacing={22}
              weekDayFormat="dd"
              withFullScreenPortal={false}
              withPortal={false}
            />
          </Col>
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col
            className="mb-1"
            sm={6}
          >
            <Button
              active={false}
              className="w-100"
              disabled={false}
              type="submit"
              variant="info"
            >
              Add Expanse 
            </Button>
          </Col>
        </FormGroup>
      </Form>
    </Col>
  </Row>
</Container>
`;

exports[`Should render ExpanseForm with error on invalid form submission 2`] = `
<Container
  fluid={false}
>
  <Row
    className="justify-content-md-center"
    noGutters={false}
  >
    <Col
      md={6}
    >
      <Alert
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="warning"
      >
        Please provide description and amount!!
      </Alert>
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Description
          </FormLabel>
          <FormControl
            id="description"
            onChange={[Function]}
            type="text"
            value=""
          />
        </FormGroup>
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Note
          </FormLabel>
          <FormControl
            as="textarea"
            id="note"
            onChange={[Function]}
            rows="2"
            value=""
          />
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col>
            <FormLabel
              column={false}
              srOnly={false}
            >
              Amount
            </FormLabel>
            <FormControl
              id="amount"
              onChange={[Function]}
              type="text"
              value=""
            />
          </Col>
          <Col
            id="singleDatePicker"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Created At
            </FormLabel>
            <br />
            <withStyles(SingleDatePicker)
              anchorDirection="right"
              appendToBody={false}
              block={false}
              calendarInfoPosition="bottom"
              customCloseIcon={null}
              customInputIcon={null}
              date={"1970-01-01T00:00:00.000Z"}
              dayPickerNavigationInlineStyles={null}
              daySize={39}
              disableScroll={false}
              disabled={false}
              displayFormat={[Function]}
              enableOutsideDays={false}
              firstDayOfWeek={null}
              focused={false}
              hideKeyboardShortcutsPanel={false}
              horizontalMargin={0}
              horizontalMonthPadding={13}
              id="date"
              initialVisibleMonth={null}
              inputIconPosition="before"
              isDayBlocked={[Function]}
              isDayHighlighted={[Function]}
              isOutsideRange={[Function]}
              isRTL={false}
              keepFocusOnInput={false}
              keepOpenOnDateSelect={false}
              monthFormat="MMMM YYYY"
              navNext={null}
              navPosition="navPositionTop"
              navPrev={null}
              noBorder={false}
              numberOfMonths={1}
              onClose={[Function]}
              onDateChange={[Function]}
              onFocusChange={[Function]}
              onNextMonthClick={[Function]}
              onPrevMonthClick={[Function]}
              openDirection="down"
              orientation="horizontal"
              phrases={
                Object {
                  "calendarLabel": "Calendar",
                  "chooseAvailableDate": [Function],
                  "clearDate": "Clear Date",
                  "closeDatePicker": "Close",
                  "dateIsSelected": [Function],
                  "dateIsUnavailable": [Function],
                  "enterKey": "Enter key",
                  "escape": "Escape key",
                  "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                  "homeEnd": "Home and end keys",
                  "jumpToNextMonth": "Move forward to switch to the next month.",
                  "jumpToPrevMonth": "Move backward to switch to the previous month.",
                  "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardShortcuts": "Keyboard Shortcuts",
                  "leftArrowRightArrow": "Right and left arrow keys",
                  "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                  "moveFocusByOneMonth": "Switch months.",
                  "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                  "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                  "openThisPanel": "Open this panel.",
                  "pageUpPageDown": "page up and page down keys",
                  "questionMark": "Question mark",
                  "returnFocusToInput": "Return to the date input field.",
                  "roleDescription": "datepicker",
                  "selectFocusedDate": "Select the date in focus.",
                  "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                  "upArrowDownArrow": "up and down arrow keys",
                }
              }
              placeholder="Date"
              readOnly={false}
              regular={false}
              renderCalendarInfo={null}
              renderDayContents={null}
              renderMonthElement={null}
              renderMonthText={null}
              renderNavNextButton={null}
              renderNavPrevButton={null}
              renderWeekHeaderElement={null}
              reopenPickerOnClearDate={false}
              required={false}
              screenReaderInputMessage=""
              showClearDate={false}
              showDefaultInputIcon={false}
              small={false}
              verticalHeight={null}
              verticalSpacing={22}
              weekDayFormat="dd"
              withFullScreenPortal={false}
              withPortal={false}
            />
          </Col>
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col
            className="mb-1"
            sm={6}
          >
            <Button
              active={false}
              className="w-100"
              disabled={false}
              type="submit"
              variant="info"
            >
              Add Expanse 
            </Button>
          </Col>
        </FormGroup>
      </Form>
    </Col>
  </Row>
</Container>
`;

exports[`Should render ExpanseForm with expanse data 1`] = `
<Container
  fluid={false}
>
  <Row
    className="justify-content-md-center"
    noGutters={false}
  >
    <Col
      md={6}
    >
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Description
          </FormLabel>
          <FormControl
            id="description"
            onChange={[Function]}
            type="text"
            value="Gas bill"
          />
        </FormGroup>
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Note
          </FormLabel>
          <FormControl
            as="textarea"
            id="note"
            onChange={[Function]}
            rows="2"
            value=""
          />
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col>
            <FormLabel
              column={false}
              srOnly={false}
            >
              Amount
            </FormLabel>
            <FormControl
              id="amount"
              onChange={[Function]}
              type="text"
              value="3000"
            />
          </Col>
          <Col
            id="singleDatePicker"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Created At
            </FormLabel>
            <br />
            <withStyles(SingleDatePicker)
              anchorDirection="right"
              appendToBody={false}
              block={false}
              calendarInfoPosition="bottom"
              customCloseIcon={null}
              customInputIcon={null}
              date={"1970-01-01T00:00:00.000Z"}
              dayPickerNavigationInlineStyles={null}
              daySize={39}
              disableScroll={false}
              disabled={false}
              displayFormat={[Function]}
              enableOutsideDays={false}
              firstDayOfWeek={null}
              focused={false}
              hideKeyboardShortcutsPanel={false}
              horizontalMargin={0}
              horizontalMonthPadding={13}
              id="date"
              initialVisibleMonth={null}
              inputIconPosition="before"
              isDayBlocked={[Function]}
              isDayHighlighted={[Function]}
              isOutsideRange={[Function]}
              isRTL={false}
              keepFocusOnInput={false}
              keepOpenOnDateSelect={false}
              monthFormat="MMMM YYYY"
              navNext={null}
              navPosition="navPositionTop"
              navPrev={null}
              noBorder={false}
              numberOfMonths={1}
              onClose={[Function]}
              onDateChange={[Function]}
              onFocusChange={[Function]}
              onNextMonthClick={[Function]}
              onPrevMonthClick={[Function]}
              openDirection="down"
              orientation="horizontal"
              phrases={
                Object {
                  "calendarLabel": "Calendar",
                  "chooseAvailableDate": [Function],
                  "clearDate": "Clear Date",
                  "closeDatePicker": "Close",
                  "dateIsSelected": [Function],
                  "dateIsUnavailable": [Function],
                  "enterKey": "Enter key",
                  "escape": "Escape key",
                  "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                  "homeEnd": "Home and end keys",
                  "jumpToNextMonth": "Move forward to switch to the next month.",
                  "jumpToPrevMonth": "Move backward to switch to the previous month.",
                  "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardShortcuts": "Keyboard Shortcuts",
                  "leftArrowRightArrow": "Right and left arrow keys",
                  "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                  "moveFocusByOneMonth": "Switch months.",
                  "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                  "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                  "openThisPanel": "Open this panel.",
                  "pageUpPageDown": "page up and page down keys",
                  "questionMark": "Question mark",
                  "returnFocusToInput": "Return to the date input field.",
                  "roleDescription": "datepicker",
                  "selectFocusedDate": "Select the date in focus.",
                  "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                  "upArrowDownArrow": "up and down arrow keys",
                }
              }
              placeholder="Date"
              readOnly={false}
              regular={false}
              renderCalendarInfo={null}
              renderDayContents={null}
              renderMonthElement={null}
              renderMonthText={null}
              renderNavNextButton={null}
              renderNavPrevButton={null}
              renderWeekHeaderElement={null}
              reopenPickerOnClearDate={false}
              required={false}
              screenReaderInputMessage=""
              showClearDate={false}
              showDefaultInputIcon={false}
              small={false}
              verticalHeight={null}
              verticalSpacing={22}
              weekDayFormat="dd"
              withFullScreenPortal={false}
              withPortal={false}
            />
          </Col>
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "displayName": "Row",
              "render": [Function],
            }
          }
        >
          <Col
            className="mb-1"
            sm={6}
          >
            <Button
              active={false}
              className="w-100"
              disabled={false}
              type="submit"
              variant="info"
            >
              Add Expanse 
            </Button>
          </Col>
          <Col
            sm={6}
          >
            <Button
              active={false}
              className="w-100"
              disabled={false}
              id="removeExpanse"
              type="button"
              variant="success"
            >
              Remove Expanse 
            </Button>
          </Col>
        </FormGroup>
      </Form>
    </Col>
  </Row>
</Container>
`;
